Index: linux-4.9.91/drivers/mtd/devices/m25p80.c
===================================================================
--- linux-4.9.91.orig/drivers/mtd/devices/m25p80.c	2018-08-14 09:17:07.917670013 +0800
+++ linux-4.9.91/drivers/mtd/devices/m25p80.c	2018-08-14 09:21:17.345668743 +0800
@@ -27,6 +27,9 @@
 #include <linux/spi/flash.h>
 #include <linux/mtd/spi-nor.h>
 
+#define OPCODE_RSTEN		0x66    /* Enable reset */
+#define OPCODE_RESET		0x99    /* Reset device */
+
 #define	MAX_CMD_SIZE		6
 struct m25p {
 	struct spi_device	*spi;
@@ -259,10 +262,25 @@
 {
 	struct m25p	*flash = spi_get_drvdata(spi);
 
+	if ((&flash->spi_nor)->addr_width > 3) {
+		printk(KERN_INFO "m25p80: exit 4-byte address mode\n");
+		flash->command[0] = SPINOR_OP_EX4B;
+		spi_write(flash->spi, flash->command, 1);
+		flash->command[0] = OPCODE_RSTEN;
+		spi_write(flash->spi, flash->command, 1);
+		flash->command[0] = OPCODE_RESET;
+		spi_write(flash->spi, flash->command, 1);
+	}
+
 	/* Clean up MTD stuff. */
 	return mtd_device_unregister(&flash->spi_nor.mtd);
 }
 
+void m25p_32m_remove(struct spi_device *spi)
+{
+	m25p_remove(spi);
+}
+
 /*
  * Do NOT add to this array without reading the following:
  *
@@ -333,6 +351,7 @@
 	.id_table	= m25p_ids,
 	.probe	= m25p_probe,
 	.remove	= m25p_remove,
+	.shutdown = m25p_32m_remove,
 
 	/* REVISIT: many of these chips have deep power-down modes, which
 	 * should clearly be entered on suspend() to minimize power use.
iff --git a/tools/firmware-utils/src/mktplinkfw.c b/tools/firmware-utils/src/mktplinkfw.c
ndex ab5fd6d..54279a4 100644
-- a/tools/firmware-utils/src/mktplinkfw.c
++ b/tools/firmware-utils/src/mktplinkfw.c
@ -162,6 +162,12 @@ static struct flash_layout layouts[] = {
		.kernel_ep	= 0xc0000000,
		.rootfs_ofs	= 0x2a0000,
	}, {
		.id             = "32Mlzma",
		.fw_max_len     = 0x1f80000,
		.kernel_la      = 0x80060000,
		.kernel_ep      = 0x80060000,
		.rootfs_ofs     = 0x140000,
	}, {
		/* terminating entry */
	}
};

